File: dsa-problems/LinkedLists/solutions/python/LNK-001-lab-roster-append.py
import sys

class ListNode:
    def __init__(self, val=0):
        self.val = val
        self.next = None

class Solution:
    def __init__(self):
        self.head = None
        self.tail = None

    def push_back(self, value: int) -> None:
        new_node = ListNode(value)
        if not self.head:
            self.head = new_node
            self.tail = new_node
        else:
            self.tail.next = new_node
            self.tail = new_node

    def to_array(self):
        result = []
        current = self.head
        while current:
            result.append(current.val)
            current = current.next
        return result

def main():
    input = sys.stdin.read
    data = input().splitlines()
    if not data:
        return
    
    n = int(data[0])
    sol = Solution()
    
    for i in range(1, n + 1):
        line = data[i].strip()
        if line.startswith("push_back"):
            _, v = line.split()
            sol.push_back(int(v))
        else:
            arr = sol.to_array()
            print(" ".join(str(x) for x in arr))

if __name__ == "__main__":
    main()

File: dsa-problems/LinkedLists/solutions/javascript/LNK-001-lab-roster-append.js
const readline = require("readline");

class ListNode {
  constructor(val) {
    this.val = val;
    this.next = null;
  }
}

class Solution {
  constructor() {
    this.head = null;
    this.tail = null;
  }

  pushBack(value) {
    const newNode = new ListNode(value);
    if (!this.head) {
      this.head = newNode;
      this.tail = newNode;
    } else {
      this.tail.next = newNode;
      this.tail = newNode;
    }
  }

  toArray() {
    const result = [];
    let current = this.head;
    while (current) {
      result.push(current.val);
      current = current.next;
    }
    return result;
  }
}

const rl = readline.createInterface({
  input: process.stdin,
  output: process.stdout,
});

let lines = [];
rl.on("line", (line) => lines.push(line.trim()));
rl.on("close", () => {
  if (lines.length === 0) return;
  const n = parseInt(lines[0], 10);
  const sol = new Solution();
  for (let i = 1; i <= n; i++) {
    const line = lines[i];
    if (line.startsWith("push_back")) {
      const parts = line.split(" ");
      sol.pushBack(parseInt(parts[1], 10));
    } else {
      const arr = sol.toArray();
      console.log(arr.join(" "));
    }
  }
});
